### ISSUE TITLE
Challenge the relevance of Redis backends in core

### ISSUE THREAD

**Issue by leplatrem** [Kinto/kinto#712]
My opinion is that having Redis in the main Kinto repo, and not in a separate repo:

* is mainly an intellectual exercise to demostrate that the backend interface is agnostic
* introduces a overhead of maintenance (redis methods to port, redis server to run the tests, redis on travis)
* has happened because Remy and Alex are huge fans of Redis :)

Whereas concretely, the Redis has a number of limitations that prevents its usage in production:

* there is no «per-request» transaction
* there is no schema migration
* the filtering and sorting is done in memory in python, after deserializing each records from json strings
* [some methods are implemented naively](https://github.com/Kinto/kinto/blob/3.2.1/kinto/core/storage/memory.py#L31-L40)

Using the redis backend for cache and postgres for storage and permissions can be relevant, but the absence of transaction (two-phase commit) prevents redis to be used along postgres for storage or permission backends.

Of course we could spend ~~some~~ a lot of time improving all this, but we don't have any use-case and it looks like the few production deployments we've heard of (mozilla, kinto-heroku etc.) are using postgresql. So we could simply move the redis backend to a separate repo :)

**What do you think ?**

![](http://i.giphy.com/Zu2mUFvWg37kA.gif)

**Comment by Natim** [comment]
I think you are missing the point of having Redis here.

We have three backends, Permissions, Cache and Storage.

The reason we have implemented the three backends with Postgresql and Redis is to prevent user from having to install both if they want to.

You are right that using Redis for the storage backend in production has got some drawbacks but using Postgresql for cache and permission is not a good fit either.

Having both is also the guarantee that the interface exposes can work with both NoSQL and SQL databases so that we can use any backends.

Most of kinto clients will never use filtering or sorting (kinto.js for instance).

You are true that we should enable transaction for the Redis storage backend and we will work on it eventually but postgresql backend didn't had transaction either at first.

The reason why we have both is not because Alex and Rémy are fans of Redis it is because Redis is currently our best option for the permission and cache backend.

**Comment by leplatrem** [comment]
> The reason we have implemented the three backends with Postgresql and Redis is to prevent user from having to install both if they want to.

*to allow users* you mean ?

> You are right that using Redis for the storage backend in production has got some drawbacks but using Postgresql for cache and permission is not a good fit either.

Why postgresql is not a good fit for cache and permission ?

> Having both is also the guarantee that the interface exposed can work with both NoSQL and SQL database

Intellectual purpose as I mentioned.

> Most of kinto clients will never use filtering or sorting (kinto.js for instance).

Not true. They use massively `?_since=<>`

> You are true that we should enable transaction for the Redis storage backend and we will work on it eventually but postgresql backend didn't had transaction either at first.

Please admit that this is a huge amount of work :)

> Redis it is because Redis is currently our best option for the permission and cache backend.

No, not without transactions :)

**Comment by Natim** [comment]
I don't mind having redis as a plugin but I would rather keep it for the same reason we have kinto-history and kinto-default-bucket as part as the core.

**Comment by Natim** [comment]
> Why postgresql is not a good fit for cache and permission ?

For cache because it doesn't expire data automatically as part as the DB feature.
For permissions because the intersection of sets is a killer feature of Redis.

I will not enter in the war of PostgreSQL is better than Redis, I don't care. For me it is important to have at least two backends implementations as a warrant of choice.

Note that we user Redis in production for the cache backend.

**Comment by tarekziade** [comment]
"a warrant of choice"

this is not an argument imho. People chose Kinto for its features not the DB it uses.

Having to deploy postgresql is just a technical detail. I would argue that if we find some speed bottleneck and Redis makes thing super efficient, let's use it. If it's just to make sure our backend can be implemented with various DBs, I would ask : why? it just sounds like more trouble & more work.

That said, for Mac Developrs, it's always a little bit painful to have to deploy postgres. It's not straightforward. So being able to play with Kinto with no DB is kind of nice. I would vote for a memory-to-disk toy implementation though :)

A cache DB should be optional btw, but it sounds like you want more from Redis - and from experience (Hello) that led us to some problems.

**Comment by Natim** [comment]
For me https://github.com/Kinto/kinto/pull/711 is a good example of what I mean by "warrant of choice", the way it has been implemented in Postgresql makes it difficult to apply to no SQL backends, we should probably add a new storage method rather than to change the existing ones so that we can do the delete all without having to return previous records or to create tombstones for instance.

What you call `memory-to-disk` is exactly what redis is offering so I would be in favor of keeping Redis rather than replacing it by another backend.

\<troll\>I don't think we can say that our experience with Hello led us to some problems, we already had more problem with Kinto and Postgresql than we had for 2 years of Hello.\</troll\>

**Comment by tarekziade** [comment]
> the way it has been implemented in Postgresql makes it difficult to apply to no SQL backends

Which leads to the question: why do you want a No SQL backend ?

> What you call memory-to-disk is exactly what redis is offering so I would be in favor of keeping Redis rather than replacing it by another backend.

What I call a memory-to-disk backend is a toy implementation in python so Kinto "works" just to try it out on a single process for fun - rather than having to install and run Redis.

> <troll>I don't think we can say that our experience with Hello led us to some problems, we already had 
more problem with Kinto and Postgresql than we had for 2 years of Hello.</troll>

*cough* *cough*   Choosing Redis over a classical database in Hello was a very poor choice. We had to deploy HUGE servers to make sure that we had enough room to fit the whole database in RAM, and then worry a lot about sharding and replication. If we had picked Postgres or MySQL from day one we woud have avoided a lot of trouble. This was just to have the TTL feature, but that cost us a lot more work than if we have had crons to clean up the DB.  We were just lucky the DB did grow more ;)

Redis is a cache server, not a real RDBMS - I think you need to admit it and stop trying to use it for everything ;)

**Comment by Natim** [comment]
For sure Redis is not a RDBMS but Kinto doesn't need one to store JSON objects.

**Comment by almet** [comment]
Hi all,

The main reason I see to having at least two backends in the core is to make sure the database APIs stay generic enough to plug new types of storages (which, is left to the community, and we do not have any community implementation so far, apart an intent from @Lothiraldan). Nevertheless, that is the rationale, and as been used as a "selling" point for Kinto -- plug with the database you want if the existing ones do not match your needs.

Now, if I read well, you would like to remove Redis from Kinto core (and put it somewhere else), in order to make things easier to develop, but I do not get how it will be easier as the codebase would still need to be maintained, unless you are thinking about removing Redis entirely from Kinto ?

Slightly related: I believe having the redis backend as an external plugin would probably make the maintenance experience harder (like what we had with cliquet + kinto), but that might just be a CI issue. In other words, if we can have a Continuous Integration system to test the plugins compatibility before merging features, that would solve this issue.

Also slightly related, but might be out of scope: the (levelup/leveldown)[https://github.com/Level/leveldown] experience might be of interest to us, and I wonder if something similar could be used in Python for Kinto, making it possible to have multiple backends maintained by *the community* (tm) and used by us for Kinto.

**Comment by leplatrem** [comment]
Yes, that's what I meant: I suggest that we move the redis backends to a separate repo, and that we stop maintaining them officially along Kinto. 

The Redis aficionados would maintain the repo at their pace, implement stuff carefully, and would have the right to complain or suggest API changes when something is not done correctly in Kinto core (signatures, changelog, ...). But currently, keeping Redis like that in core does not guarantee anything and is counterproductive — since methods are implemented poorly — and making it suitable for production requires a lot of time/energy — since it has no migrations, no transactions, a lot of re-entrancies, is sub-optimized. 

An investment in kinto core that I personally don't see worth making right now since Redis is not the recommended stack for production.

> Note: In parallel, I opened a few questions for the [next kinto-monthly meeting](https://pad.notmyidea.org/p/kinto-monthly).

**Comment by tarekziade** [comment]
> I suggest that we move the redis backends to a separate repo, and that we stop maintaining them officially along Kinto. 

+1

**Comment by Natim** [comment]
While implementing Quotas I realized that both MemoryBackend and RedisBackend are currently unable to abort a transaction in the `ResourceChanged` event.

Both the kinto-signer and the quota plugin are using this transaction aborting feature to make the request fails in case of 5XX errors.

The redis backend is not suitable in its current state for use with these plugins.
Let's move the backend to a external repository and let the community to support it.

**Comment by Natim** [comment]
The Redis backends moved to https://github.com/Kinto/kinto-redis
